packages/cdk8s-plus-17/src/job.ts

```diff
@@ -13,6 +13,40 @@ import { Volume } from './volume';
  */
 export interface JobProps extends ResourceProps, PodTemplateProps {
 
+  /**
+   * Specifies the duration in seconds relative to the startTime that
+   * the job may be active before the system tries to terminate it.
+   *
+   * @default - If unset, then there is no deadline.
+   */
+  readonly activeDeadline?: Duration;
+
+  /**
+   * Specifies the number of retries before marking this job failed.
+   *
+   * @default - If not set, system defaults to 6.
+   */
+  readonly backoffLimit?: number;
+
+  /**
+   * Specifies the desired number of successfully finished pods the job should be run with.
+   * Setting to 1 means that parallelism is limited to 1
```

Comment:
I'm not sure I understand this. Does this also mean that setting it to 2 will limit parallelism to 2? 
Intuitively i imagined that it must be that `spec.completions <= spec.parallelism`

Reply:
This part always seemed a bit confusing to me as well.  I am simply copying the documentation for the properties of completions and parallelism from https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#jobspec-v1-batch.  I think the best reference is what they refer to from there about https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/

Reply:
@abierbaum Yeah its definitely tricky. But I think this requires some more attention.

One of the biggest benefits of `cdk8s-plus` is the ability to reduce cognitive load, i.e simplify and clarify confusing properties and features. It's up to us as authors to fully understand how everything works and come up with better modeling. This presents a great opportunity to do that.

The [documentation](https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/) you pointed to can be codified, making it so users don't have to go through it to figure out what configuration should be applied to get their desired outcome.

 For example, it states that:

> There are three main types of task suitable to run as a Job:
> 1. Non-parallel Jobs
> 2. Parallel Jobs with a fixed completion count
> 3. Parallel Jobs with a work queue

So we can think of something like:

```ts
// default - non parallel single completion. doesn't offer 'completions` or `parallelism`
new kplus.Job(this, 'Job');

// fixed completion, offers a mandatory completion count, and an optional parallelism that default to 1 and doesn't
// allow a number greater than count
new kplus.FixedCompletionJob(this, 'Job', {
  count: 5
})

// work queue, offers only parallelism as per the docs: 
// For a work queue Job, you must leave .spec.completions unset, and set .spec.parallelism to a non-negative integer.
new kplus.WorkQueueJob(this, 'Job', {
  parallelism: 5
})
```

This can also look something like:

```ts
new kplus.Job(this, 'Job', {
  execution: Execution.fixedCompletion(...) // or Execution.workQueue(...) 
})
```

WDYT? Would you be up for creating a separate PR for the parallelism stuff where we can discuss the right approach?

I'm good with merging the other two properties since they seem to be really straight forward.
 
Also note that since we now have [escape hatches](https://cdk8s.io/docs/v1.0.0-beta.3/concepts/escape-hatches/), while not super elegant, users will always be able to explicitly set any property they want if we haven't yet come up with the right API for it.

Reply:
@iliapolo I can do that.  May be a couple days.

As to the escape hatches, one question there.  As I used cdk8s+ more, I see places where other properties are missing (for example `Deployment` doesn't expose `strategy` and `revisionHistoryLimit`).  These are to important ones for what I am doing with a deployment.  But what is the general policy about what part of the spec's should get exposed and not?

I understand the idea of reducing cognitive load, and agree that it is a great goal.  cdk8s+ is massively helpful to reduct the cognitive load of selectors and connecting things up like services to deployments. 

I just don't know where a good line is for what is exposed because if something doesn't get exposed to pass to the apiObject spec below it, then suddenly the cognitive load massively increases because people have to start thinking about JSON patching instead of object/interface properties.

How are you thinking about that part?  It would help me understand where to submit PRs for general use or just use json patch for my local use case.

Reply:
@abierbaum Eventually we'd like cdk8s+ to cover the entire surface area, so there is no line with regards to what should or shouldn't be exposed. 

The only thing to consider is whether the API **can** offer a better experience than just passing through the properties, and if it can, we should hold back until we come up with that API.

Sometimes that won't be necessary, like for the `activeDeadLine` property, and sometimes it won't be possible. But we should ask ourself that question for every property, and whether or not to submit a PR will depend on the answer.

If you find yourself unsure, just open an issue where we can discuss the API prior to diving into the code.